Added _very_ necessary kernel scaling to the kernel_centerer and clea…

rosecers opened the pull request:
…ned up a few areas

The previous iteration of kernel centering didn't include scaling by n_train / trace(K)

I went through, added the scaling, and cleaned up a few things.

hurricane642 commented on skcosmo/preprocessing/flexible_scaler.py:
```diff
@@ -114,15 +114,9 @@ class KernelFlexibleCenterer(TransformerMixin, BaseEstimator):
-    :type ref_cmean: array of shape (1, n_features)
-
-    :param ref_mean: an average for the whole kernel matrix
-    :type ref_mean: array
```
Why just delete a piece of documentation? These are important parameters, and you just remove them.

rosecers replied:
They're parameters for the _fit_ function, not the overall class. This isn't an appropriate place to have them.

hurricane642 commented on skcosmo/preprocessing/flexible_scaler.py:
```diff
@@ -114,15 +114,9 @@ class KernelFlexibleCenterer(TransformerMixin, BaseEstimator):
     """
 
-    def __init__(self, ref_cmean=None, ref_mean=None):
+    def __init__(self):
```
No, you need to keep the old `__init__.` It makes checking everything much easier.

rosecers replied:
You _don't_ initialize the scaler with the fitting parameters, you fit it with them. Just like when setting up a LR or PCA, you don't initialize with the fitting parameters.

hurricane642 commented on skcosmo/preprocessing/flexible_scaler.py:
```diff
@@ -175,13 +178,12 @@ def transform(self, K, y=None):
-        for key in parameters.keys():
-            if parameters[key] is None:
+        for key in ["ref_cmean", "ref_mean", "scale", "reference_shape"]:
+            if not hasattr(self, key) or getattr(self, key) is None:
```
I do not agree with this listing at all. First, such code makes it difficult for any kind of code extensibility (if needed). Secondly, it overloads the code, and it is absolutely unclear why.

rosecers replied:
So the previous instantiation had a few flaws -- if there were additional class attributes, then the loop would cycle through them and check. Additionally, if none of these were initialized, then this test would totally miss the error. It needs to be explicit.

hurricane642 commented on skcosmo/preprocessing/flexible_scaler.py:
```diff
@@ -198,7 +200,7 @@ def transform(self, K, y=None):
             - rmean.reshape((K.shape[0], 1))
             + np.broadcast_arrays(K, self.ref_mean)[1]
-        )
+        ) / self.scale
```
Why we need to add `self.scale` here?

rosecers replied:
We need to scale the kernel to appropriate size, see the appendix of the KPCovR paper for the derivation.

hurricane642 replied:
Ok, thank you!

hurricane642 commented on tests/kernel_flexible_centerer.py:
```diff
@@ -38,13 +38,15 @@ def test_fit_transform(self):
         K = np.random.uniform(0, 100, size=(3, 3))
         model = KernelFlexibleCenterer()
-        Ktr = model.fit_transform(K)
+        Ktr = model.fit_transform(K.copy())
```
Why use `.copy` here?

rosecers replied:
eh, this isn't totally necessary, just a preference when working with modifying functions.

hurricane642 replied:
Well, why not, ok :).

hurricane642 commented on tests/kernel_flexible_centerer.py:
```diff
@@ -54,16 +56,18 @@ def test_fit_reference(self):
         K = np.random.uniform(0, 100, size=(3, 3))
-        ref_cmean = np.random.uniform(0, 100, size=(3))
-        ref_mean = np.random.uniform(0, 100, size=(1))[0]
+
```
In general, this test is written in order to check whether we are working correctly with the reference parameters or not. Now it just duplicates `test_fit_transform`, why is it needed? And yes, now we do not check the work on reference parameters, which is generally important.

rosecers replied:
Good point -- I'll revert the meaning of this test.

hurricane642 replied:
Ok, great!

hurricane642 reviewed (changes requested):
I tried to describe in as much detail as possible what I don't like.

ceriottm commented:
Re parameters-in-init vs parameters-in-fit, can you check what sklearn
does? This is not an "absolute truth" kind of scenario, these are two
options and we should aim for consistency.

On Mon, 7 Dec 2020 at 13:02, rosecers <notifications@github.com> wrote:

> *@rosecers* commented on this pull request.
> ------------------------------
>
> In tests/kernel_flexible_centerer.py
> <https://github.com/cosmo-epfl/scikit-cosmo/pull/38#discussion_r537452897>
> :
>
> > @@ -38,13 +38,15 @@ def test_fit_transform(self):
>          """
>          K = np.random.uniform(0, 100, size=(3, 3))
>          model = KernelFlexibleCenterer()
> -        Ktr = model.fit_transform(K)
> +        Ktr = model.fit_transform(K.copy())
>
> eh, this isn't totally necessary, just a preference when working with
> modifying functions.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cosmo-epfl/scikit-cosmo/pull/38#discussion_r537452897>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAIREZ3Z5FRYIFNVOHWFLYLSTS745ANCNFSM4UQJOIGA>
> .
>

rosecers commented:
> Re parameters-in-init vs parameters-in-fit, can you check what sklearn does? This is not an "absolute truth" kind of scenario, these are two options and we should aim for consistency.
> […](#)
> On Mon, 7 Dec 2020 at 13:02, rosecers ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In tests/kernel_flexible_centerer.py <[#38 (comment)](https://github.com/cosmo-epfl/scikit-cosmo/pull/38#discussion_r537452897)> : > @@ -38,13 +38,15 @@ def test_fit_transform(self): """ K = np.random.uniform(0, 100, size=(3, 3)) model = KernelFlexibleCenterer() - Ktr = model.fit_transform(K) + Ktr = model.fit_transform(K.copy()) eh, this isn't totally necessary, just a preference when working with modifying functions. — You are receiving this because you are subscribed to this thread. Reply to this email directly, view it on GitHub <[#38 (comment)](https://github.com/cosmo-epfl/scikit-cosmo/pull/38#discussion_r537452897)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAIREZ3Z5FRYIFNVOHWFLYLSTS745ANCNFSM4UQJOIGA> .

Sklearn does parameters in fit in both of their scalers/centerers

hurricane642 commented:
Great, then the question about documentation disappears. But to be honest, I still didn't understand, why do we need to do a `fit` check in the new format? I understand that this is how we get rid of unnecessary code in `__init__`, but I think the updated version is a bit overloaded. What's the problem with the previous version? And yes, thank you very much for going through the code again ^_^.

rosecers commented:
> Great, then the question about documentation disappears. But to be honest, I still didn't understand, why do we need to do a `fit` check in the new format? I understand that this is how we get rid of unnecessary code in `__init__`, but I think the updated version is a bit overloaded and more simple to initialize parameters in the `__init__`. Or it is not a `sklearn` style? What's the problem with the previous version? And yes, thank you very much for going through the code again ^_^.

So if you look in a lot of the sklearn scripts, they explicitly note which parameters need to be present in order to call a model fitted. In general, specificity is king with these sorts of checks. If you want an example, look at `sklearn/utils/estimator_checks.py` L 2184-2186

hurricane642 commented:
Okay, fair enough :( Well then everything is fine, I agree with everything :).

rosecers commented:
@hurricane642 one more round of edits -- I changed the internal names to all end in an underscore, as is common with sklearn parameters, and renamed ref_cmean and ref_mean to K_fit_rows and K_fit_all, which albeit a little weird, are the names of these parameters in sklearn.

hurricane642 commented:
I agree, it looks really strange, but yes, the truth is in the style of `sklearn`. Going back to the previous topic, I looked at `KernelCenterer` - they use a special function `check_is_fitted`. Maybe we should use this function in our code too?

rosecers commented:
> I agree, it looks really strange, but yes, the truth is in the style of `sklearn`. Going back to the previous topic, I looked at `KernelCenterer` - they use a special function `check_is_fitted`. Maybe we should use this function in our code too?

good point -- it looks like it should be an easy change, let me see

hurricane642 commented:
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py, line 977

rosecers commented:
> https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py, line 977

Yep! Go take a look

hurricane642 commented:
Seems really great, I like it!
